code/iLasso_for_SCINGE.m

```diff
@@ -20,47 +20,101 @@
 %
 % MIT License
 % Copyright (c) 2014 USC-Melady
-addDC = 0;
-BIC = 0;
+pa = m.pa;
 BIC_bias = 0;
 AIC_bias = 0;
 % Parameters
 L = L/dDt;     % Length of studied lag
 L0 = L;
 Dt = dDt;
 % \Delta t
-%SIG = 1;    % Kernel parameter. Here Gaussian Kernel Bandwidth
-B = sum(Series{1}(2, :)<=(L*Dt));
-N1 = size(Series{1}, 2);
-P = length(Series);
+[LX,WX] = size(m,'X');
+if ismember('regix',who(m))
+    numregs = length(m,'regix');
+else
+    numregs = LX;
+end
+% Define funtion for Gaussian kernel
+gkern = @(x,y) gausskernel(x,y,SIG);
+% Load expression matrix and pseudotime
+X = m.X;
+ptime = m.ptime;
+
+% Precompute the full kernel matrix -- to be subsampled for each regulator
+if m.computeKp
+    if ismember('fullKp',who(m))&&~isempty(m.fullKp)
+        m.fullKp(1,:) = [];
+    end
+    Kp2.Kp = (zeros(length(ptime)));
+    refT = (ptime);
+    for k = 1:L
+        Kp = (refT-(L)*Dt+(k-1)*Dt);
+        Kp = bsxfun(gkern,Kp',refT);
+        Kp2(k).Kp = (Kp);
+        m.fullKp(1,k) = Kp2(k);
+    end
+    m.computeKp = 0;
+    clear Kp Kp2;
+end
+
+X1 = m.X1;
+tind = find(X1(pa(1),:));
+tval = X(pa(1),:);
+%target{1}(:,tind) = [];
+ttime = ptime;
+ttime(tind) = [];
+tval(tind) = [];
+B = sum(ttime<=(L*Dt));
+N1 = size(ttime, 2);
 
 % Build the matrix elements
-Am = (zeros(N1-B, P*L));
-bm = (zeros(N1-B, 1));
+Am = (zeros(N1-B, numregs*L));
 % Building the design matrix
-for j = 1:P
-    for i = (B+1):N1
-        bm(i-B) = Series{1}(1, i);
-        ti = (Series{1}(2, i) - (L)*Dt):Dt:(Series{1}(2, i)-Dt);
-   %     ti = repmat(ti, length(Series{j}(2, :)), 1);
- %       tSelect = repmat(Series{j}(2, :)', 1, L0);
-        tSelect = Series{j}(2, :)';
-        %ySelect = repmat(Series{j}(1, :)', 1, L0);
-        ySelect = Series{j}(1, :)';
+for k = 1:L
+    Kp2 = m.fullKp(1,k);
+    for j = 1:numregs
+        %tSelect = repmat(Series{j}(2, :)',1,L0);
+        rind = find(X1(pa(j),:));
+        ySelect = (full(X(pa(j),:)'));
+        refL = length(ySelect);
+        tSelect = ptime;
+        ySelect(rind) = [];
+        tSelect(rind) = [];
+        refind = 1:refL;
+        remind = refind(~(ismember(refind,tind))); % setdiff(refind,tind);
+        remind = union(1:remind(B+1)-1,tind);
+        rind = refind(~(ismember(refind,rind)));
+        remind = refind(~(ismember(refind,remind)));
+        %Kp = Kp2(1).Kp;
+        %    for i = (B+1):N1
+        %bm(i-B) = Series{1}(1, i);
+        % ti = ones(length(Series{j}(2, :)), 1)*((Series{1}(2, i) - (L)*Dt):Dt:(Series{1}(2, i)-Dt+Dt*params.AllowZeroLag));
+        %ti = repmat((Series{1}(2, i) - (L)*Dt):Dt:(Series{1}(2, i)-Dt+Dt*params.AllowZeroLag),length(tSelect),1);
+        %     Kp = (ttime(1, B+1:end)' - (L)*Dt)+(k-1)*Dt;
         switch krnl
```

Comment:
Do we still support multiple types of kernels?

Reply:
Actually, we don't. Should I remove that as part of the next set of optimizations?

Reply:
Yes, please remove these lines for other kernels and remove the `krnl` function argument in the next round of updates.